File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/control/localization/EulerIntegration.java
package org.firstinspires.ftc.teamcode.control.localization;

import org.firstinspires.ftc.teamcode.util.Point;
import org.firstinspires.ftc.teamcode.util.Pose;

import static org.firstinspires.ftc.teamcode.util.MathUtil.epsilonEquals;

public class EulerIntegration extends BaseOdometry {

    public EulerIntegration(Pose startPose) {
        super(startPose);
    }

    @Override
    protected void robotPoseUpdate() {
        double dtheta = currPoseDelta.heading;
        double sineTerm, cosTerm;

        if (epsilonEquals(dtheta, 0)) {
            sineTerm = 1.0 - dtheta * dtheta / 6.0;
            cosTerm = dtheta / 2.0;
        } else {
            sineTerm = Math.sin(dtheta) / dtheta;
            cosTerm = (1 - Math.cos(dtheta)) / dtheta;
        }

        Point fieldTranslationDelta = new Point(
                sineTerm * currPoseDelta.x - cosTerm * currPoseDelta.y,
                cosTerm * currPoseDelta.x + sineTerm * currPoseDelta.y
        );

        Pose fieldPoseDelta = new Pose(fieldTranslationDelta.rotated(currPose.heading), dtheta);
        currPose.add(fieldPoseDelta);
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/control/localization/EulerIntegration.kt
package org.firstinspires.ftc.teamcode.control.localization

import org.firstinspires.ftc.teamcode.util.math.Point
import kotlin.math.absoluteValue
import kotlin.math.cos
import kotlin.math.sin

object EulerIntegration {
    fun update(dx: Double, lDelta: Double, rDelta: Double, dh: Double): Point {
        return if (dh.absoluteValue > 0) {
            val radiusOfMovement = (lDelta + rDelta) / (2 * dh)
            val radiusOfStrafe = dx / dh

            Point(
                radiusOfMovement * (1 - cos(dh)) + (radiusOfStrafe * sin(dh)),
                (radiusOfMovement * sin(dh)) + (radiusOfStrafe * (1 - cos(dh)))
            )
        } else Point(dx, (lDelta - rDelta) / 2.0)
    }
}
